### ISSUE TITLE
[Bug] Four live aiv_shard results produce incorrect output in a bidirectional A2/A3 kernel

### ISSUE THREAD

**Issue by lyfne123** [hw-native-sys/pypto-lib#1194]
## Background

While migrating Qwen3-14B paged attention from explicit C/V synchronization to `pl.aiv_shard` / `pl.aic_gather`, a small mixed cube/vector kernel produces incorrect output on `a2a3`. Preserving numerical correctness is a prerequisite for comparing performance against the existing `decode_fwd` implementation. The reproducer below is independent of Qwen model code and uses constant, exactly representable values.

| Component | Reproduced version |
|---|---|
| pypto-lib | `0e7e07aa5e42db6e08c68fd5bfe1def571492b53` (`main`, with local Qwen experiments; full reproducer below) |
| PyPTO | `9f657f37ed20ce148b46fb7229c267a152a0644e` (`main-test`, editable installation) |
| simpler runtime | `4e4d3a4ad1e54c1db3d50e72decc025a9075bfa0` (detached, matches PyPTO gitlink) |
| PTOAS | `0.57` (matches `toolchain/versions.env`) |
| PTO ISA | `a8040450238f162985d8b596fbebeb54bfba2bf5` (detached, managed checkout matches runtime's `pto_isa.pin`) |
| CANN | `9.0.0` |
| Python / torch | `3.10.19` / `2.6.0+cpu` |
| Host / target | Linux aarch64 / `a2a3` |

The runtime source checkout and toolchain pins are consistent. The installed runtime binary's build revision was not independently stamped/verified.

### Diagnosis

Ownership is not yet isolated across PyPTO lowering, PTOAS, and PTO ISA, so this is filed in pypto-lib for routing. **Incorrect output is confirmed; overwrite of live receive slots is only a hypothesis.** The generated code uses a bidirectional pipe with two slots and emits four C2V pops before the first concat/free. This evidence alone does not establish which component violates the transfer/lifetime contract.

### Description

The kernel computes four QK products, receives all four results through `aiv_shard`, concatenates them, casts to BF16, gathers them back to AIC, multiplies by V, and shards the final result for storage. There are no loop-carried shard/gather results, softmax approximations, cross-half GM exchange, or model dependencies.

With Q and V filled with ones and the four W pages filled with 1, 2, 3, and 4, the intermediate page values should be 128, 256, 384, and 512. Every final element should equal `128 * (128 + 256 + 384 + 512) = 163840`. The observed first row instead contains `196608` (20% too large).

### Steps to Reproduce

1. Use the pinned environment above and source the CANN environment. Set `PTOAS_ROOT` to the installed PTOAS 0.57 release root (containing its `ptoas.sh` launcher).
2. Save the complete source below as `repro_shard_roundtrip.py` in the pypto-lib checkout root. Its only repository import is the public `golden` harness at the listed pypto-lib commit.
3. Allocate one A2/A3 device through the normal device allocator and run:

```bash
PYTHONPATH=. python repro_shard_roundtrip.py --device <allocated-device>
```

The captured failure was run through `task-submit --device auto` with PTOAS 0.57. No fixed device index is required.

<details>
<summary>Complete reproducer: <code>repro_shard_roundtrip.py</code> (save locally)</summary>

```python
import argparse
import os
import torch
import pypto.language as pl
from golden import TensorSpec, run_jit

@pl.jit
def four_live_shards(q: pl.Tensor[[16, 128], pl.BF16], w: pl.Tensor[[512, 128], pl.BF16],
                     v: pl.Tensor[[512, 128], pl.BF16], out: pl.Out[pl.Tensor[[16, 128], pl.FP32]]):
    with pl.at(level=pl.Level.CORE_GROUP):
        qt = pl.load(q, [0, 0], [16, 128], target_memory=pl.MemorySpace.Mat)
        w0 = pl.load(w, [0, 0], [128, 128], target_memory=pl.MemorySpace.Mat)
        w1 = pl.load(w, [128, 0], [128, 128], target_memory=pl.MemorySpace.Mat)
        w2 = pl.load(w, [256, 0], [128, 128], target_memory=pl.MemorySpace.Mat)
        w3 = pl.load(w, [384, 0], [128, 128], target_memory=pl.MemorySpace.Mat)
        c0 = pl.matmul(qt, pl.tile.transpose_view(w0), out_dtype=pl.FP32)
        c1 = pl.matmul(qt, pl.tile.transpose_view(w1), out_dtype=pl.FP32)
        c2 = pl.matmul(qt, pl.tile.transpose_view(w2), out_dtype=pl.FP32)
        c3 = pl.matmul(qt, pl.tile.transpose_view(w3), out_dtype=pl.FP32)
        for lane in pl.split_aiv(2, mode=pl.SplitMode.UP_DOWN):
            s0 = pl.aiv_shard(c0)
            s1 = pl.aiv_shard(c1)
            s2 = pl.aiv_shard(c2)
            s3 = pl.aiv_shard(c3)
            s01 = pl.concat(s0, s1)
            s23 = pl.concat(s2, s3)
            result = pl.concat(s01, s23)
            probability = pl.cast(result, target_type=pl.BF16)
            joined = pl.aic_gather(probability)
        vt = pl.load(v, [0, 0], [512, 128], target_memory=pl.MemorySpace.Mat)
        acc = pl.matmul(joined, vt, out_dtype=pl.FP32)
        for lane in pl.split_aiv(2, mode=pl.SplitMode.UP_DOWN):
            shard = pl.aiv_shard(acc)
            pl.store(shard, [lane * 8, 0], out)
    return out

def golden(values):
    values['out'][:] = (values['q'].float() @ values['w'].float().T).bfloat16().float() @ values['v'].float()

def compare(actual, expected, **kwargs):
    print('ACTUAL row0:', actual[0, :16].tolist(), flush=True)
    print('EXPECTED row0:', expected[0, :16].tolist(), flush=True)
    return torch.equal(actual, expected), 'four live C2V shards must retain distinct values'

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--device', type=int, default=int(os.environ.get('TASK_DEVICE', '0')))
    args = parser.parse_args()
    specs = [
        TensorSpec('q', [16, 128], torch.bfloat16, init_value=lambda: torch.ones(16, 128, dtype=torch.bfloat16)),
        TensorSpec('w', [512, 128], torch.bfloat16, init_value=lambda: torch.arange(1, 5, dtype=torch.bfloat16).repeat_interleave(128).reshape(512, 1).expand(512, 128).contiguous()),
        TensorSpec('v', [512, 128], torch.bfloat16, init_value=lambda: torch.ones(512, 128, dtype=torch.bfloat16)),
        TensorSpec('out', [16, 128], torch.float32, is_output=True),
    ]
    result = run_jit(fn=four_live_shards, specs=specs, golden_fn=golden,
                     compare_fn={'out': compare}, compile_cfg={'dump_passes': True},
                     runtime_cfg={'platform': 'a2a3', 'device_id': args.device})
    raise SystemExit(0 if result.passed else 1)
```

</details>

### Expected Behavior

The output must equal the BF16-roundtripped torch reference exactly for these inputs. If keeping four received tiles live exceeds a supported transfer contract, compilation should reject the source with an actionable diagnostic instead of silently returning incorrect values.

### Actual Behavior

The harness fails exact comparison. First 16 elements of row 0:

```text
ACTUAL:   [196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0, 196608.0]
EXPECTED: [163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0, 163840.0]
```

Generated PTO excerpt, with source-location annotations removed:

```mlir
  pto.aiv_initialize_pipe {dir_mask = 3, slot_size = 16384, slot_num = 2}
      (gm_slot_buffer = %arg4 : !pto.ptr<f32>, c2v_consumer_buf = %four_live_shards_incore_0_c2v_slot_buffer : i32, v2c_consumer_buf = %four_live_shards_incore_0_v2c_slot_buffer_import : i32)
  %s0__ssa_v0 = pto.tpop_from_aic {split = 1} -> !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  %s1__ssa_v0 = pto.tpop_from_aic {split = 1} -> !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  %s2__ssa_v0 = pto.tpop_from_aic {split = 1} -> !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  %s3__ssa_v0 = pto.tpop_from_aic {split = 1} -> !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  %s01__ssa_v0 = pto.alloc_tile addr = %c32768_i64 valid_row = %c8_index valid_col = %c256_index : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=256, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  pto.tconcat ins(%s0__ssa_v0, %s1__ssa_v0 : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%s01__ssa_v0 : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=256, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
  pto.tfree_from_aic {split = 1}
  pto.tfree_from_aic {split = 1}
  %s23__ssa_v0 = pto.alloc_tile addr = %c40960_i64 valid_row = %c8_index valid_col = %c256_index : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=256, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>
  pto.tconcat ins(%s2__ssa_v0, %s3__ssa_v0 : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=128, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%s23__ssa_v0 : !pto.tile_buf<loc=vec, dtype=f32, rows=8, cols=256, v_row=?, v_col=?, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
  pto.tfree_from_aic {split = 1}
  pto.tfree_from_aic {split = 1}
```

### Environment

See the exact versions and pin-consistency notes in Background. The numerical failure requires device execution; it is not a compile-only result.

### Host Platform

Linux (aarch64)

### Additional Context

- In the Qwen candidate, consuming/concatenating received shards in pairs before receiving the next pair allowed the tested paged-attention case matrix and single-layer decode golden checks to pass. This is an application workaround, not proof of a general upstream fix or of the precise cause of this micro-reproducer's failure.
- A smaller C2V-only control passed, but it changed both transfer direction and tile width; it does not isolate direction as the cause. A copy-per-shard variant has not been validated.
- Related historical issues: [PyPTO #2269](https://github.com/hw-native-sys/pypto/issues/2269) (bidirectional GM workspace sizing) and [#2287](https://github.com/hw-native-sys/pypto/issues/2287) (consumer ring reserve sizing). Those are closed; equivalence to their causes has not been established on this pinned environment.
- [PyPTO #1981](https://github.com/hw-native-sys/pypto/issues/1981) involved cross-half GM exchange across repeated model dispatches; this reproducer has neither that exchange nor a multi-layer dispatch loop.
